architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,379 @@
+        "_EXPERIMENTAL_FEATURES_ENABLED": [
+          "Extensibility"
+        ],
+        "_generator": {
+          "name": "bicep",
+          "version": "0.29.47.4906",
+          "templateHash": "14156924711842038952"
+        }
+      },
+      "parameters": {
+        "application": {
+          "type": "string"
+        },
+        "tag": {
+          "type": "string"
+        },
+        "port": {
+          "type": "int"
+        }
+      },
+      "imports": {
+        "Radius": {
+          "provider": "Radius",
+          "version": "latest"
+        }
+      },
+      "resources": {
+        "container": {
+          "import": "Radius",
+          "type": "Applications.Core/containers@2023-10-01-preview",
+          "properties": {
+            "name": "container",
+            "properties": {
+              "application": "[parameters('application')]",
+              "container": {
+                "image": "[format('ghcr.io/radius-project/samples/demo:{0}', parameters('tag'))]",
+                "ports": {
+                  "web": {
+                    "containerPort": "[parameters('port')]"
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  parameters: |
+    {
+      "application": {
+        "value": ""
+      },
+      "tag": {
+        "value": "latest"
+      },
+      "port": {
+        "value": 3000
+      }
+    }
+```
+
+## Design
+
+### High Level Design
+
+This design proposes the creation of two new Kubernetes custom resources and their corresponding reconcilers: `BicepDeployment` and `BicepResource`. The `BicepDeployment` controller will be responsible for deploying and reconciling resources defined in ARM JSON manifests via Radius, while the `BicepResource` controller will be responsible for reconciling individual resources. The `BicepDeployment` controller will have a control loop that will check the status of an in-progress operation, process deletion, and process creation or update. The `BicepResource` controller will have a similar control loop that will check the status of an in-progress operation or process deletion if necessary.
+
+### Architecture Diagram
+
+![Architecture Diagram](2024-08-bicepdeployment-controller/architecture.png)
+
+### Detailed Design
+
+#### BicepDeployment Custom Resource Definition
+
+The `BicepDeployment` CRD will be a new Kubernetes CRD that will be used to contain the ARM JSON manifest and parameters. The CRD will have the following fields:
+
+```go
+// BicepDeploymentSpec defines the desired state of a BicepDeployment
+type BicepDeploymentSpec struct {
+  // Template is the ARM JSON manifest that defines the resources to deploy.
+  Template string `json:"template"`
+
+  // Parameters is the ARM JSON parameters for the template.
+  Parameters string `json:"parameters"`
+}
```

Comment:
We need to include the provider configuration here as well. This is where the scopes for providers like `radius`, `azure`, `aws` get passed in.

Reply:
Take a look at existing CLI code where we do rad deploy
\